runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java

```diff
@@ -103,43 +156,76 @@ public void process(ProcessContext c) {
+    public PTransformReplacement<PCollection<KV<K, V>>, PCollection<KV<ShardedKey<K>, Iterable<V>>>>
         getReplacementTransform(
             AppliedPTransform<
-                    PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>, GroupIntoBatches<K, V>>
+                    PCollection<KV<K, V>>,
+                    PCollection<KV<ShardedKey<K>, Iterable<V>>>,
+                    GroupIntoBatches<K, V>.WithShardedKey>
                 transform) {
       return PTransformReplacement.of(
           PTransformReplacements.getSingletonMainInput(transform),
-          new StreamingGroupIntoBatches(runner, transform.getTransform()));
+          new StreamingGroupIntoBatchesWithShardedKey<>(runner, transform.getTransform()));
     }
 
     @Override
     public Map<PCollection<?>, ReplacementOutput> mapOutputs(
-        Map<TupleTag<?>, PCollection<?>> outputs, PCollection<KV<K, Iterable<V>>> newOutput) {
+        Map<TupleTag<?>, PCollection<?>> outputs,
+        PCollection<KV<ShardedKey<K>, Iterable<V>>> newOutput) {
       return ReplacementOutputs.singleton(outputs, newOutput);
     }
   }
 
   /**
-   * Specialized implementation of {@link GroupIntoBatches} for unbounded Dataflow pipelines. The
-   * override does the same thing as the original transform but additionally record the input to add
-   * corresponding properties during the graph translation.
+   * Specialized implementation of {@link GroupIntoBatches.WithShardedKey} for unbounded Dataflow
+   * pipelines. The override does the same thing as the original transform but additionally records
+   * the input of {@code GroupIntoBatchesDoFn} in order to append relevant step properties during
+   * the graph translation.
    */
-  static class StreamingGroupIntoBatches<K, V>
-      extends PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>> {
+  static class StreamingGroupIntoBatchesWithShardedKey<K, V>
+      extends PTransform<PCollection<KV<K, V>>, PCollection<KV<ShardedKey<K>, Iterable<V>>>> {
 
     private final transient DataflowRunner runner;
-    private final GroupIntoBatches<K, V> original;
+    private final GroupIntoBatches<K, V>.WithShardedKey original;
 
-    public StreamingGroupIntoBatches(DataflowRunner runner, GroupIntoBatches<K, V> original) {
+    public StreamingGroupIntoBatchesWithShardedKey(
+        DataflowRunner runner, GroupIntoBatches<K, V>.WithShardedKey original) {
       this.runner = runner;
       this.original = original;
     }
 
     @Override
-    public PCollection<KV<K, Iterable<V>>> expand(PCollection<KV<K, V>> input) {
-      runner.maybeRecordPCollectionWithAutoSharding(input);
-      return input.apply(original);
+    public PCollection<KV<ShardedKey<K>, Iterable<V>>> expand(PCollection<KV<K, V>> input) {
+      PCollection<KV<ShardedKey<K>, V>> intermediate_input = ShardKeys(input);
+
+      runner.maybeRecordPCollectionWithAutoSharding(intermediate_input);
+
+      if (original.getMaxBufferingDuration() != null) {
```

Comment:
This doesn't look like it'll scale if more options are used. Why not just apply original?

Reply:
We need to recognize the GroupIntoBatchesDoFn which is a private member of the transform. Here we record the input pcoll of GroupIntoBatchesDoFn, which would be the output of the key-sharding DoFn, so in the translation we can append autosharding properties for those steps whose input was recorded. This doesn't scale indeed. Any suggestions?

Reply:
Sounds like privacy is getting in the way of writing clean code here. What about making it package private, and adding a (package) helper to identify it?

Reply:
Tried to make the changes but seems like making the `GroupIntoBatchesDoFn` visible to runners is not enough, since now we need to differentiate the DoFn in `GroupIntoBatches` transform and in `GroupIntoBatches.withShardedKey` transform. `DataflowPipelineTranslator` seems to only recognize primitive transforms, e.g., `ParDoSingle` and `ParDo.MultiOutput`.

https://github.com/apache/beam/blob/f67cb9ad6a46fb5b6af9aababc25305a8d5968c1/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L997

Possible workarounds:
- Match the DoFn name (withShardedKey) or PCollection type name (ShardedKey).
- Record the output pcollection of GIB transform and use that to recognize GIB DoFn. Not 100% sure if that will work but I can try.
- Or...?

Reply:
Also, the original now shard the input based on thread ids. For the override, we probably don't want to pre-shard the input in SDK.

Reply:
The latest commit I pushed took the second approach above. In the current version we record the output PCollection of `GroupIntoBatches.WithShardedKey` when replacing the original transform and add necessary properties to the DoFn that produces the recorded PCollection. That way we don't need to replicate the implementation but instead simply apply the original transform as is.